### PR TITLE
set cluster/route/listener/vhost name len based on CLI opt

### PR DESCRIPTION
**Pull request by rshriram**
Automatically set the max allowed length for cluster, route_config_name and listener name based on the max_stat_name_len specified. The name len is max_stat_name_len - 67 (which is the suffix reserved for internally generated stats).

Signed-off-by: Shriram Rajagopalan <shriram@us.ibm.com>

cc @hennna / @ggreenway FYI.

@mattklein123 this is per discussion on Slack.

### CODE REVIEW

**Comment by mattklein123** [comment]
Quick pre-review comments: Will need doc changes. Also, we should do this for RDS and LDS also.

File: include/envoy/server/options.h [hunk 1]
```diff
@@ -130,6 +130,11 @@ class Options {
    * @return uint64_t the maximum name length of a stat.
    */
   virtual uint64_t maxStatNameLength() PURE;
+
+  /**
+   * @return uint64_t the maximum name length of a cluster/route/listener.
+   */
+  virtual uint64_t maxUserSuppliedNameLength() PURE;
```
Comment (ggreenway) [inline include/envoy/server/options.h]: I think that it would be better to have this value be derived from the maxStatNameLen, or possibly have that one be derived from this one.  Being able to specify both seems confusing.

**Comment by htuch** [comment]
@hennna please review this from the Google side (I will do the final pass from us).

File: include/envoy/server/options.h (hunk 1 above)
Comment (mattklein123) [reply include/envoy/server/options.h]: Yup exactly. No more options are needed. Please rework as @ggreenway suggests.

Comment (rshriram) [reply include/envoy/server/options.h]: Yeah good point. Does it suffice to just derive max cluster name length as max_stat_name_len - reserved_stat_suffix_len (which happens to be 67 at the moment) ? That would make things considerably simple.

Comment (mattklein123) [reply include/envoy/server/options.h]: Yup that's exactly what I would do.

File: source/common/stats/stats_impl.h [hunk 2]
```diff
@@ -103,6 +111,8 @@ struct RawStatData {
 
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
+  // total length must be equal to DEFAULT_MAX_NAME_SIZE
+  static const size_t DEFAULT_MAX_STAT_SUFFIX_LENGTH = 67;
```
Comment (ggreenway) [inline source/common/stats/stats_impl.h]: rename to MAX_STAT_SUFFIX_LENGTH?  Since it isn't configurable, I don't think of it as a default; it's the only value.

Comment (rshriram) [reply source/common/stats/stats_impl.h]: done..

**Comment by rshriram** [comment]
@ggreenway I too feel that its probably better to have the identity name be the configurable option and derive the other one. @htuch / @mattklein123 LMK if you want to replace the max-stat-name-len with max-user-field-name-len or something.. 

Btw, its not  just cluster. Its cluster name, route_config_name, listener name.

**Comment by mattklein123** [comment]
> Given that we just added the command-line option a couple weeks ago, @mattklein123 what do you think about replacing it with one for max user supplied name len, and computing the max name len from that?

That's fine. Since we haven't release 1.5.0 yet we can change it. Please just make the defaults such that we don't need to do a full restart (the shared memory version should be the same).

**Comment by rshriram** [comment]
@mattklein123 / @ggreenway PTAL.. changed the CLI option and compile time option as well.

File: source/common/config/utility.cc [hunk 3]
```diff
@@ -119,5 +123,13 @@ std::string Utility::resourceName(const ProtobufWkt::Any& resource) {
       fmt::format("Unknown type URL {} in DiscoveryResponse", resource.type_url()));
 }
 
+void Utility::checkUserSuppliedNameLength(const std::string& error_prefix,
+                                          const std::string& name) {
+  if (name.length() > Stats::RawStatData::maxUserSuppliedNameLength()) {
+    throw EnvoyException(fmt::format("{}: name exceeds allowed maximum {}", error_prefix,
+                                     Stats::RawStatData::maxUserSuppliedNameLength()));
```
Comment (ggreenway) [inline source/common/config/utility.cc]: Please also print the passed in name, and possibly the length, so it's easier to figure out what was too big and by how much it was too big.

File: source/common/stats/stats_impl.h [hunk 4]
```diff
@@ -103,6 +109,8 @@ struct RawStatData {
 
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
+  // total length must be equal to DEFAULT_MAX_NAME_SIZE
+  static const size_t MAX_STAT_SUFFIX_LENGTH = 67;
```
Comment (ggreenway) [inline source/common/stats/stats_impl.h]: Please only define this value in one place.  It is duplicated in options_impl.cc right now.

File: source/server/options_impl.cc [hunk 5]
```diff
@@ -144,6 +150,6 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
   drain_time_ = std::chrono::seconds(drain_time_s.getValue());
   parent_shutdown_time_ = std::chrono::seconds(parent_shutdown_time_s.getValue());
   max_stats_ = max_stats.getValue();
-  max_stat_name_length_ = max_stat_name_len.getValue();
+  max_stat_name_length_ = max_obj_name_len.getValue() + ENVOY_MAX_STAT_SUFFIX_LENGTH;
```
Comment (ggreenway) [inline source/server/options_impl.cc]: The user-specified option is now max_stat_name_len.  For consistency, please replace the method  Options::maxStatNameLength() with Options::maxObjNameLen()

Comment (rshriram) [reply source/server/options_impl.cc]: I intentionally left it as it is. Replacing it with maxObjectNameLen will require each call site to use maxObjectNameLen + maxStatSuffixLen. That seems unnecessary detail across the code base where maxStatNameLen is used.

File: source/common/config/utility.cc (hunk 3 above)
Comment (rshriram) [reply source/common/config/utility.cc]: I can, but be warned that its of little use. The fields are named "name" which is not of much use. Thats the reason I opted to use the error_prefix and have some meaningful error message as part of the error.

File: source/server/options_impl.cc (hunk 5 above)
Comment (ggreenway) [reply source/server/options_impl.cc]: There should only end up being one call-site: RawStatData::configure().  And then there will be 2 static functions on RawStatData; maxNameLength() and maxObjNameLength(), one of which will be implemented in terms of the other (adding or subtracting 67).

File: test/server/hot_restart_impl_test.cc [hunk 6]
```diff
@@ -132,15 +136,15 @@ TEST_P(HotRestartImplAlignmentTest, objectOverlap) {
   std::vector<TestStat> stats;
   for (uint64_t i = 0; i < num_stats_; i++) {
     std::string name = fmt::format("{}zzzzzzzzzzzzzzzzzzzzzzzzzzzz", i)
-                           .substr(0, Stats::RawStatData::maxNameLength());
+                           .substr(0, Stats::RawStatData::maxObjNameLength());
```
Comment (ggreenway) [inline test/server/hot_restart_impl_test.cc]: I think this should stay maxNameLength(), because it's directly creating the stat name

File: test/server/hot_restart_impl_test.cc [hunk 7]
```diff
@@ -132,15 +136,15 @@ TEST_P(HotRestartImplAlignmentTest, objectOverlap) {
...
     TestStat ts;
     ts.stat_ = hot_restart_->alloc(name);
     ts.name_ = ts.stat_->name_;
     ts.index_ = i;
 
     // If this isn't true then the hard coded part of the name isn't long enough to make the test
     // valid.
-    EXPECT_EQ(ts.name_.size(), Stats::RawStatData::maxNameLength());
+    EXPECT_EQ(ts.name_.size(), Stats::RawStatData::maxObjNameLength());
```
Comment (ggreenway) [inline test/server/hot_restart_impl_test.cc]: Same here: maxNameLength()

File: test/server/hot_restart_impl_test.cc [hunk 8]
```diff
@@ -50,8 +52,10 @@ class HotRestartImplTest : public testing::Test {
 
 TEST_F(HotRestartImplTest, versionString) {
   setup();
-  EXPECT_EQ(hot_restart_->version(), Envoy::Server::SharedMemory::version(
-                                         options_.maxStats(), options_.maxStatNameLength()));
+  EXPECT_EQ(hot_restart_->version(),
+            Envoy::Server::SharedMemory::version(options_.maxStats(),
+                                                 options_.maxObjNameLength() +
```
Comment (ggreenway) [inline test/server/hot_restart_impl_test.cc]: I think this is correct, but as per @mattklein123's request, can you make sure the version string is the same before and after your change (as a manual check)?

Comment (rshriram) [reply test/server/hot_restart_impl_test.cc]: with the new build its `9.200.16384.127` and with master build, its `9.200.16384.127` . so it matches!

File: bazel/README.md [hunk 9]
```diff
@@ -245,11 +245,12 @@ on the Bazel command line.
 
 ## Stats Tunables
 
-The default maximum number of stats in shared memory, and the default maximum length of a stat name,
-can be overriden at compile-time by defining `ENVOY_DEFAULT_MAX_STATS` and `ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH`,
+The default maximum number of stats in shared memory, and the default maximum length of a cluster/
```
Comment (mattklein123) [inline bazel/README.md]: nit: reflow text.

File: source/common/config/json_utility.h [hunk 10]
```diff
@@ -6,6 +6,12 @@
 
 // NOLINT(namespace-envoy)
 
+// Set a string field in a protobuf message with the corresponding string value
+#define JSON_UTIL_SET_STRING_VALUE(json, message, field_name, value)                               \
+  do {                                                                                             \
+    (message).set_##field_name(value);                                                             \
```
Comment (mattklein123) [inline source/common/config/json_utility.h]: What does this have to do with JSON? Why is this macro needed at all?

File: source/server/options_impl.cc [hunk 11]
```diff
@@ -82,9 +83,12 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
                                       "Maximum number of stats guages and counters "
                                       "that can be allocated in shared memory.",
                                       false, ENVOY_DEFAULT_MAX_STATS, "uint64_t", cmd);
-  TCLAP::ValueArg<uint64_t> max_stat_name_len("", "max-stat-name-len",
-                                              "Maximum name length for a stat", false,
-                                              ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH, "uint64_t", cmd);
+  TCLAP::ValueArg<uint64_t> max_obj_name_len("", "max-obj-name-len",
```
Comment (mattklein123) [inline source/server/options_impl.cc]: needs doc update

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 12]
```diff
@@ -242,10 +242,7 @@ TEST_F(ClusterManagerImplTest, MaxClusterName) {
   }
   )EOF";
 
-  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromJson(json)), Json::Exception,
-                            "JSON at lines 4-6 does not conform to schema.\n Invalid schema: "
-                            "#/properties/name\n Schema violation: maxLength\n Offending "
-                            "document key: #/name");
+  EXPECT_THROW(create(parseBootstrapFromJson(json)), EnvoyException);
```
Comment (mattklein123) [inline test/common/upstream/cluster_manager_impl_test.cc]: Why does this throw now? What's the new error?

File: source/common/stats/stats_impl.h [hunk 13]
```diff
@@ -103,6 +115,8 @@ struct RawStatData {
 
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
```
Comment (hennna) [inline source/common/stats/stats_impl.h]: Should this be DEFAULT_MAX_OBJ_NAME_SIZE? Wondering if it would be better to have all max name size parameters be derived.

File: source/server/options_impl.cc [hunk 14]
```diff
@@ -18,12 +19,12 @@
...
-#ifndef ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH
-#define ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH 127
+#ifndef ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH
+#define ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH 60
 #endif
 
-#if ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH < 127
-#error "ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH must be >= 127"
+#if ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH < 60
+#error "ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH must be >= 60"
```
Comment (hennna) [inline source/server/options_impl.cc]: Outside the scope of this PR... Will there come a time when we also want to have MAX_STAT_SUFFIX_LENGTH be configurable? Or is that better left hard coded and bumped up when necessary?

**Comment by rshriram** [comment]
On Tue, Oct 17, 2017 at 11:08 PM hennna <notifications@github.com> wrote:

> *@hennna* commented on this pull request.
>
> Thanks! Looking forward to using this feature.
> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145274398>:
>
> > @@ -103,6 +115,8 @@ struct RawStatData {
>
>  private:
>    static const size_t DEFAULT_MAX_NAME_SIZE = 127;
> +  // total length must be equal to DEFAULT_MAX_NAME_SIZE
> +  static const size_t MAX_STAT_SUFFIX_LENGTH = 67;
>
> Maybe document where this number comes from? It seems from #667
> <https://github.com/envoyproxy/envoy/pull/667> it was from a cluster
> stats name plus some buffer.
> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145306685>:
>
> > @@ -68,6 +68,18 @@ struct RawStatData {
>    }
>
>    /**
> +   * Returns the maximum length of a user supplied object (route/cluster/listener)
> +   * name field in a stat. This length does not include a trailing NULL-terminator.
> +   */
> +  static size_t maxObjNameLength() { return maxNameLength() - MAX_STAT_SUFFIX_LENGTH; }
>
> Should we instead have maxNameLength be derived from maxObjNameLength?
>

We could. I was trying to avoid touching the functions that were playing
tricks with static vars, as the maxNameLen function is used everywhere! But
let me see if it’s easy to change.

> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145306841>:
>
> > @@ -103,6 +115,8 @@ struct RawStatData {
>
>  private:
>    static const size_t DEFAULT_MAX_NAME_SIZE = 127;
>
> Should this be DEFAULT_MAX_OBJ_NAME_SIZE? Wondering if it would be better
> to have all max name size parameters be derived.
>

This is max length of a stat entry. I agree it’s worded in a bit confusing
manner.

> ------------------------------
>
> In source/server/options_impl.cc
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145307198>:
>
> >  #endif
>
> -#if ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH < 127
> -#error "ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH must be >= 127"
> +#if ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH < 60
> +#error "ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH must be >= 60"
>
> Outside the scope of this PR... Will there come a time when we also want
> to have MAX_STAT_SUFFIX_LENGTH be configurable? Or is that better left hard
> coded and bumped up when necessary?
>

@mattklein123 says we can bump it when necessary. I am surprised that 67
characters are enough to hold our current stat names.

File: source/server/options_impl.cc (hunk 11 above)
Comment (rshriram) [reply source/server/options_impl.cc]: ack. Sorry missed that totally.

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 12 above)
Comment (rshriram) [reply test/common/upstream/cluster_manager_impl_test.cc]: This was always throwing an error. Earlier, it threw a JSON schema error with some message like "Invalid #/properties/name...." .. I merely changed this to use EnvoyException, as we are no longer validating the length using JSON schema.

In other words, UX wise, you would still get an exception when max length is exceeded, except it will be worded differently.

**Comment by rshriram** [comment]
@mattklein123 / @junr03 I noticed this in the docs: 

```Virtual cluster statistics are output in the
*vhost.<virtual host name>.vcluster.<virtual cluster name>.* namespace and include the following
statistics:
```

Given that this PR allows for configurable name lengths for clusters, how does it affect the above stat, which is concatenating two different user supplied identities?

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 12 above)
Comment (mattklein123) [reply test/common/upstream/cluster_manager_impl_test.cc]: Right, please keep EXPECT_THROW_WITH_MESSAGE and switch it to the new error.

**Comment by mattklein123** [comment]
> Given that this PR allows for configurable name lengths for clusters, how does it affect the above stat, which is concatenating two different user supplied identities?

This occurred to me also. I think there are probably a bunch of random places where we should be guarding against user supplied stat names, etc. that might be too large. I would probably just open up a follow up ticket to audit.

**Comment by rshriram** [comment]
@mattklein123 I moved all the objNameLength checks into a separate file. Also added name length checks for virtual host (but not the virtual host + virtual cluster). 
@hennna addressed your comments as well. PTAL

File: docs/operations/cli.rst [hunk 15]
```diff
@@ -117,10 +117,19 @@ following are the command line options that Envoy supports.
   during a hot restart. See the :ref:`hot restart overview <arch_overview_hot_restart>` for more
   information. Defaults to 900 seconds (15 minutes).
 
-.. option:: --max-stat-name-len <uint64_t>
+.. _operations_cli_max_obj_name_len:
```
Comment (mattklein123) [inline docs/operations/cli.rst]: doc nit: You don't need this anchor, you can reference options directly and then they look nice inline. See how we ref other options.

File: bazel/README.md [hunk 16]
```diff
@@ -245,11 +245,14 @@ on the Bazel command line.
 
 ## Stats Tunables
 
-The default maximum number of stats in shared memory, and the default maximum length of a stat name,
-can be overriden at compile-time by defining `ENVOY_DEFAULT_MAX_STATS` and `ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH`,
-respectively, to the desired value.  For example:
+The default maximum number of stats in shared memory, and the default
+maximum length of a cluster/ route config/listener name, can be
```
Comment (mattklein123) [inline bazel/README.md]: nit: extra space in "cluster/ "

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 17]
```diff
@@ -232,22 +232,6 @@ TEST_F(ClusterManagerImplTest, UnknownHcType) {
...
-  const std::string json = R"EOF(
-  {
-    "clusters": [
-    {
-      "name": "clusterwithareallyreallylongnamemorethanmaxcharsallowedbyschema"
-    }]
-  }
-  )EOF";
-
-  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromJson(json)), Json::Exception,
```
Comment (mattklein123) [inline test/common/upstream/cluster_manager_impl_test.cc]: Isn't this test still valid? Or is it not anymore?

File: test/common/router/rds_impl_test.cc [hunk 18]
```diff
@@ -182,6 +182,27 @@ TEST_F(RdsImplTest, UnknownCluster) {
...
+    },
+    "codec_type": "auto",
+    "stat_prefix": "foo",
+    "filters": [
+      { "name": "http_dynamo_filter", "config": {} }
+    ]
+  }
+  )EOF";
+
+  EXPECT_THROW(RouteConfigProviderUtil::create(parseHttpConnectionManagerFromJson(config_json),
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: Can we do EXPECT_THROW_WITH_MESSAGE here?

Comment (rshriram) [reply test/common/router/rds_impl_test.cc]: Actually, can I get rid of it? We have the same test in utility_test.cc

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 17 above)
Comment (rshriram) [reply test/common/upstream/cluster_manager_impl_test.cc]: Same comment as above. Test just moved to utility_test.cc ..

File: test/common/router/rds_impl_test.cc (hunk 18 above)
Comment (mattklein123) [reply test/common/router/rds_impl_test.cc]: Yup that's fine

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 17 above)
Comment (mattklein123) [reply test/common/upstream/cluster_manager_impl_test.cc]: Yup sounds good

File: source/server/options_impl.cc [hunk 19]
```diff
@@ -93,14 +99,16 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
...
+  if (max_obj_name_len.getValue() < 60) {
+    std::cerr << "error: the 'max-obj-name-len' value specified (" << max_obj_name_len.getValue()
+              << ") is less than the minimum value of 60" << std::endl;
     exit(1);
   }
 
   if (hot_restart_version_option.getValue()) {
-    std::cerr << hot_restart_version_cb(max_stats.getValue(), max_stat_name_len.getValue());
+    std::cerr << hot_restart_version_cb(max_stats.getValue(),
+                                        max_obj_name_len.getValue() +
```
Comment (ggreenway) [inline source/server/options_impl.cc]: It would be nicer if we could call RawStatData::configure(), then use RawStatData::maxNameLength() here, instead of doing the addition.  But I think it's probably more effort than it's worth, because the options_impl doesn't have it's values yet at this point in the function, so probably fine to leave as is, and the tests should catch if this gets messed up somehow.

**Comment by hennna** [comment]
LGTM!
